### PR TITLE
Implement the Fetch method

### LINKED ISSUE #11898: Implement the fetch method
Specification: https://fetch.spec.whatwg.org/#fetch-method

This is assigned to our RGSoC team, so please talk with me before starting to work on any part of it.
Blocks #11894.

### PR DESCRIPTION
**Pull request by jeenalee**
This PR implements the fetch method, which is described in [the Fetch Spec](https://fetch.spec.whatwg.org/#fetch-method). The expected wpt results are updated as well.

A few comments about the current fetch implementation:
- The fetch method manually calls `JSAutoCompartment` in order to prevent SpiderMonkey from crashing. This may have to change in the future.
- Not all `FetchResponseListener` methods are implemented.
- `net_traits::CoreResourceMsg::Fetch` message takes a `net_traits::request::RequestInit`. However, when the fetch method is called, a `RequestBinding::RequestInit` is given. The fetch method constructs a `dom::request::Request` with the given `RequestInit`, then creates `net_traits::request::RequestInit` from the dom Request object for the fetch message.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11898 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13323)

### CODE REVIEW

File: components/script/dom/headers.rs [hunk 1]
```diff
@@ -217,6 +217,10 @@ impl Headers {
         self.guard.set(new_guard)
     }
 
+    pub fn set_guard_immutable(&self) {
```
Comment (jdm) [inline components/script/dom/headers.rs]: We can use `set_guard` instead.

File: components/script/dom/response.rs [hunk 2]
```diff
@@ -288,3 +290,20 @@ impl ResponseMethods for Response {
 fn serialize_without_fragment(url: &Url) -> &str {
     &url[..Position::AfterQuery]
 }
+
+impl Response {
+    pub fn set_headers(&self, option_hyper_headers: Option<Serde<HyperHeaders>>) {
+        self.Headers().set_headers( match option_hyper_headers {
```
Comment (jdm) [inline components/script/dom/response.rs]: nit: extra space after (

File: components/script/fetch.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+use js::jsapi::JSAutoCompartment;
+use net_traits::{FetchResponseListener, Metadata, NetworkError};
+use net_traits::CoreResourceMsg::Fetch as NetTraitsFetch;
+use net_traits::request::{Request as NetTraitsRequest, RequestInit as NetTraitsRequestInit};
+use network_listener::{NetworkListener, PreInvoke};
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+use url::Url;
+
+#[allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/fetch.rs]: Do you remember why this was required?

File: components/script/fetch.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+struct FetchContext {
+    fetch_promise: Option<TrustedPromise>,
+    response_object: Trusted<Response>,
+}
+
+fn from_referer_to_referer_url(request: &NetTraitsRequest) -> Option<Url> {
+    let referer = request.referrer.borrow();
```
Comment (jdm) [inline components/script/fetch.rs]: `referrer`

File: components/script/fetch.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+use std::sync::{Arc, Mutex};
+use url::Url;
+
+#[allow(unrooted_must_root)]
+struct FetchContext {
+    fetch_promise: Option<TrustedPromise>,
+    response_object: Trusted<Response>,
+}
+
+fn from_referer_to_referer_url(request: &NetTraitsRequest) -> Option<Url> {
```
Comment (jdm) [inline components/script/fetch.rs]: `from_referrer_to_referrer_url`

File: components/script/fetch.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    let referer = request.referrer.borrow();
+    match referer.to_url() {
```
Comment (jdm) [inline components/script/fetch.rs]: `referrer.to_url().map(|url| url.clone())`

File: components/script/fetch.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    let response = Response::new(global);
+
+    // Step 3
+    response.Headers().set_guard_immutable();
+
+    // Step 2
+    let request = match Request::Constructor(global, input, init) {
+        Err(e) => {
+            promise.maybe_reject_error(promise.global().r().get_cx(),
+                                       e);
```
Comment (jdm) [inline components/script/fetch.rs]: nit: this can join the previous line.

File: components/script/dom/webidls/Fetch.webidl [hunk 8]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// https://fetch.spec.whatwg.org/#fetch-method
+
+[Exposed=(Window,Worker)]
+
+partial interface WindowOrWorkerGlobalScope {
+  [NewObject, Throws] Promise<Response> fetch(RequestInfo input, optional RequestInit init);
```
Comment (jdm) [inline components/script/dom/webidls/Fetch.webidl]: No need for `Throws` anymore.

File: components/script/fetch.rs [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        Ok(r) => r.get_request(),
+    };
+
+    let request_init = request_init_from_request(request);
+
+    // Step 4
+    let (action_sender, action_receiver) = ipc::channel().unwrap();
+    let fetch_context = Arc::new(Mutex::new(FetchContext {
+        fetch_promise: Some(TrustedPromise::new(promise.clone())),
+        response_object: Trusted::new(response.r()),
```
Comment (jdm) [inline components/script/fetch.rs]: We can use `&*` instead of `.r()` here.

File: components/script/fetch.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    }));
+    let listener = NetworkListener {
+        context: fetch_context.clone(),
+        // TODO: double check how to get script_chan
```
Comment (jdm) [inline components/script/fetch.rs]: This comment is meaningless here; let's remove it.

File: components/script/fetch.rs [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    let request_init = request_init_from_request(request);
+
+    // Step 4
+    let (action_sender, action_receiver) = ipc::channel().unwrap();
+    let fetch_context = Arc::new(Mutex::new(FetchContext {
+        fetch_promise: Some(TrustedPromise::new(promise.clone())),
+        response_object: Trusted::new(response.r()),
+    }));
+    let listener = NetworkListener {
+        context: fetch_context.clone(),
```
Comment (jdm) [inline components/script/fetch.rs]: I don't think this clone is necessary.

File: components/script/fetch.rs [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        // TODO
+    }
+
+    fn process_request_eof(&mut self) {
+        // TODO
+    }
+
+    #[allow(unrooted_must_root)]
+    fn process_response(&mut self, metadata: Result<Metadata, NetworkError>) {
+        let promise = match self.fetch_promise.take() {
```
Comment (jdm) [inline components/script/fetch.rs]: Let's use `let promise = self.fetch_promise.take().expect("fetch promise is missing").root();`

**Review by jdm — CHANGES_REQUESTED** [review]
Nice work!

File: components/script/fetch.rs (hunk 3 above)
Comment (jeenalee) [inline components/script/fetch.rs]: It was indeed not required. Thank you for pointing it out!

**Comment by jeenalee** [comment]
@jdm Review comments addressed!

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Great! You can squash your commits together, and then we'll merge this once #12830 is merged.

**Comment by jeenalee** [comment]
@jdm: Should I squash all of the three commits into one?

**Comment by jdm** [comment]
Yep.

**Comment by jdm** [comment]
@bors-servo: delegate+
This can merge whenever it's rebased.

**Comment by jeenalee** [comment]
@jdm: I updated the code to reflect the changes in net_traits/lib.rs. Can you take a look at the changes when you have a moment? Thank you!

File: components/script/fetch.rs [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+                        unsafe_: _,
+                    } => {
+                        self.response_object.root().set_headers(metadata.headers);
+                        self.response_object.root().set_raw_status(metadata.status);
+                        self.response_object.root().set_final_url(metadata.final_url);
+                    },
+                    FetchMetadata::Filtered {
+                        filtered: FilteredMetadata::Opaque,
+                        unsafe_: _,
+                    } => self.response_object.root().set_type(DOMResponseType::Opaque),
```
Comment (KiChjang) [inline components/script/fetch.rs]: I prefer this style more:
```rust
FetchMetadata::Filtered { filtered, .. } => match filtered {
    FilteredMetadata::Transparent(metadata) => {
        self.response_object.root().set_headers(metadata.headers);
        // ... and others, could probably extract the 3 lines into a helper method
    }
    FilteredMetadata::Opaque => self.response_object.root().set_type(DOMResponseType::Opaque)
}
```

File: components/script/fetch.rs [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    fn process_request_body(&mut self) {
+        // TODO
+    }
+
+    fn process_request_eof(&mut self) {
+        // TODO
+    }
+
+    #[allow(unrooted_must_root)]
+    fn process_response(&mut self, fetchmetadata: Result<FetchMetadata, NetworkError>) {
```
Comment (KiChjang) [inline components/script/fetch.rs]: nit: Snake case instead (`fetch_metadata`).

File: components/script/fetch.rs [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+            // Step 4.1
+            Err(_) => {
+                promise.reject_error(
+                    promise.global().r().get_cx(),
+                    Error::Type("Network error occurred".to_string()));
+                self.fetch_promise = Some(TrustedPromise::new(promise));
+                return;
+            },
+            // Step 4.2
+            Ok(fetchmetadata_) => {
```
Comment (KiChjang) [inline components/script/fetch.rs]: Perhaps call this `metadata`?

File: components/script/fetch.rs [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+                match fetchmetadata_ {
+                    FetchMetadata::Unfiltered(metadata) => {
+                        self.response_object.root().set_headers(metadata.headers);
+                        self.response_object.root().set_raw_status(metadata.status);
+                        self.response_object.root().set_final_url(metadata.final_url);
+                    },
+                    FetchMetadata::Filtered {
+                        filtered: FilteredMetadata::Transparent(metadata),
```
Comment (KiChjang) [inline components/script/fetch.rs]: And this just `meta`, `data`, or maybe even `m`?

**Review by KiChjang — CHANGES_REQUESTED** [review]
Minor changes required, but looks good otherwise!

File: components/script/fetch.rs (hunk 13 above)
Comment (jeenalee) [inline components/script/fetch.rs]: That is a lot more elegant. Thanks!

**Comment by jeenalee** [comment]
@KiChjang Addressed your comments. Thank you!

**Review by KiChjang — APPROVED** [review]

**Comment by jdm** [comment]
Lots of test expectation updates required :)

**Comment by jeenalee** [comment]
@jdm Test expectations updated!

**Comment by jdm** [comment]
@bors-servo: retry
* weird cargo failure

**Comment by jdm** [comment]
@bors-servo: retry
* weird cargo git failure

**Comment by KiChjang** [comment]
@bors-servo retry
* Cargo is a git

**Comment by jdm** [comment]
Looks like a few final test result changes.

**Comment by jeenalee** [comment]
@jdm tests updated!

**Comment by KiChjang** [comment]
```
Ran 7578 tests finished in 1040.0 seconds.
  • 7575 ran as expected. 1677 tests skipped.
  • 3 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect-preflight.html:
  │ TIMEOUT [expected FAIL] Redirect 303: cors to same origin (preflight after redirection success case)
  └   → Test timed out

  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect-preflight.html:
  └ NOTRUN [expected TIMEOUT] Redirect 303: cors to same origin (preflight after redirection failure case)

  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect-credentials.html:
  │ TIMEOUT [expected PASS] Redirect 307 from remote to another remote with password
  └   → Test timed out

  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect-credentials.html:
  └ NOTRUN [expected PASS] Redirect 308 from same origin to remote with user and password

  ▶ Unexpected subtest result in /fetch/api/cors/cors-redirect-credentials.html:
  └ NOTRUN [expected TIMEOUT] Redirect 308 from same origin to remote with user

  ▶ Unexpected subtest result in /fetch/api/redirect/redirect-location-worker.html:
  │ TIMEOUT [expected PASS] Redirect 302 in "follow" mode with invalid location
  └   → Test timed out

  ▶ Unexpected subtest result in /fetch/api/redirect/redirect-location-worker.html:
  └ NOTRUN [expected TIMEOUT] Redirect 302 in "manual" mode with invalid location
```

**Comment by jeenalee** [comment]
@jdm: I disabled the `/redirect` and `/cors` tests.

**Comment by KiChjang** [comment]
@bors-servo retry
* wpt wat

File: tests/wpt/include.ini [hunk 17]
```diff
@@ -19,6 +19,12 @@ skip: true
   skip: false
 [fetch]
   skip: false
+  [api]
+    skip: false
+      [cors]
```
Comment (jdm) [inline tests/wpt/include.ini]: I think this line (and the subsequent ones) are over-indented, which is throwing off the manifest parser.
```
[api]
  skip: false
  [cors]
    skip: true
  [redirect]
    skip: true
```

Comment (jeenalee) [inline tests/wpt/include.ini]: Fixed!

**Comment by jdm** [comment]
@bors-servo: retry
* #13458

**Comment by jdm** [comment]
@bors-servo: retry
* #13463

**Comment by jdm** [comment]
@bors-servo: retry
* #13458

**Comment by jdm** [comment]
@bors-servo: retry
* #13341, #13482

**Comment by jdm** [comment]
@bors-servo: retry
* #13492

**Comment by jeenalee** [comment]
I re-introduced the change ([disable request-cache.html](https://github.com/servo/servo/commit/8a15e78)) that was lost during rebase.

**Comment by jdm** [comment]
@bors-servo: retry
#13359

**Comment by jeenalee** [comment]
Rebased as #13345 was merged (hooray!)!

**Comment by jdm** [comment]
@jeenalee: Does this include new test results for all the tests? Working body methods will have changed the results from previous runs significantly.

**Comment by jeenalee** [comment]
@jdm I built and ran `fetch/api/` tests, and updated the test results. Will that suffice?

**Comment by jdm** [comment]
@bors-servo: retry
#12654

**Pull request merged**